### ISSUE TITLE
Unable to create Metal library on Intel MBP, macOS 10.15

### ISSUE THREAD

**Issue by HarukaMa** [hashcat/hashcat#3882]
**GitHub is for bugs and features - not for support**
For support, please use the hashcat forums https://hashcat.net/forum/

**Check the FAQ**
Some items that might appear to be issues are not issues. Please review the hashcat FAQ https://hashcat.net/wiki/doku.php?id=frequently_asked_questions before submitting a bug report.

**Describe the bug**
Hashcat fails to create Metal library on my 2019 Intel MacBook Pro, with error message:

```
hc_mtlCreateLibraryWithSource(): failed to create metal library, Compilation failed:

program_source:7:10: fatal error: '/usr/local/share/hashcat/OpenCL/inc_vendor.h' file not found
#include M2S(INCLUDE_PATH/inc_vendor.h)
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<built-in>:2:16: note: expanded from here
#define M2S(x) XM2S(x)
               ^~~~~~
<built-in>:15:17: note: expanded from here
#define XM2S(x) #x
                ^
<scratch space>:2:1: note: expanded from here
"/usr/local/share/hashcat/OpenCL/inc_vendor.h"
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


* Device #1: Kernel /usr/local/share/hashcat/OpenCL/shared.cl build failed.
```

Strangely, the file exists:
```
$ head /usr/local/share/hashcat/OpenCL/inc_vendor.h
/**
 * Author......: See docs/credits.txt
 * License.....: MIT
 */

#ifndef INC_VENDOR_H
#define INC_VENDOR_H

#if defined HC_CPU_OPENCL_EMU_H
#define IS_NATIVE
```

I've tried to hardcode the path for the includes, but still got the same error.

I've tried to run hashcat in the build dir (non-installed version), and the include path correctly expands to the source dir, but still returns with the same error.

I suspect it has something to do with Catalina as it is an old system, but can't test it on other machines right now.

**To Reproduce**
On specified machine / OS, run `hashcat -m 100 -b -d 1`.

**Expected behavior**
There should be no compile errors.

**Hardware/Compute device (please complete the following information):**
- Compute device name: Any
- OpenCL/CUDA driver name: N/A
- OpenCL/CUDA driver version: N/A
- OpenCL/CUDA driver source: N/A
- Metal version: 212.8

**Hashcat version (please complete the following information):**
 - OS: macOS
 - Distribution: Catalina 10.15.7
 - Version: master (v6.2.6-775-ga46c9cb81+)

**Diagnostic output compute devices:**

```
For NV: Post nvidia-smi output. This tool also exist on Windows
For AMD ROCm: Post rocm-smi and rocminfo output

N/A
```

<details>
<summary> `clinfo` </summary>

```
Number of platforms                               1
  Platform Name                                   Apple
  Platform Vendor                                 Apple
  Platform Version                                OpenCL 1.2 (Jun  8 2020 17:36:15)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event

  Platform Name                                   Apple
Number of devices                                 3
  Device Name                                     Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Device Vendor                                   Intel
  Device Vendor ID                                0xffffffff
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.1
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     CPU
  Device Profile                                  FULL_PROFILE
  Max compute units                               16
  Max clock frequency                             2400MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             1024x1x1
  Max work group size                             1024
  Preferred work group size multiple              1
  Preferred / native vector sizes
    char                                                16 / 16
    short                                                8 / 8
    int                                                  4 / 4
    long                                                 2 / 2
    half                                                 0 / 0        (n/a)
    float                                                4 / 4
    double                                               2 / 2        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    64, Little-Endian
  Global memory size                              68719476736 (64GiB)
  Error Correction support                        No
  Max memory allocation                           17179869184 (16GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        64
  Global Memory cache line                        16777216 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   1 bytes
    Pitch alignment for 2D image buffers          1 bytes
    Max 2D image size                             8192x8192 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Global
  Local memory size                               32768 (32KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     4096 (4KiB)
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      1ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_APPLE_fp64_basic_ops cl_APPLE_fixed_alpha_channel_orders cl_APPLE_biased_fixed_point_image_formats cl_APPLE_command_queue_priority

  Device Name                                     Intel(R) UHD Graphics 630
  Device Vendor                                   Intel Inc.
  Device Vendor ID                                0x1024500
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.2(Oct 29 2020 23:08:45)
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Max compute units                               24
  Max clock frequency                             1250MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Preferred work group size multiple              32
  Preferred / native vector sizes
    char                                                 1 / 1
    short                                                1 / 1
    int                                                  1 / 1
    long                                                 1 / 1
    half                                                 0 / 0        (n/a)
    float                                                1 / 1
    double                                               0 / 0        (n/a)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (n/a)
  Address bits                                    64, Little-Endian
  Global memory size                              1610612736 (1.5GiB)
  Error Correction support                        No
  Max memory allocation                           402653184 (384MiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            25165824 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   4 bytes
    Pitch alignment for 2D image buffers          32 bytes
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      80ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_image2d_from_buffer cl_khr_gl_depth_images cl_khr_depth_images cl_khr_3d_image_writes

  Device Name                                     AMD Radeon Pro 5500M Compute Engine
  Device Vendor                                   AMD
  Device Vendor ID                                0x1021e00
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.2 (Oct 29 2020 23:01:05)
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Max compute units                               24
  Max clock frequency                             1450MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Preferred work group size multiple              32
  Preferred / native vector sizes
    char                                                 4 / 4
    short                                                2 / 2
    int                                                  1 / 1
    long                                                 1 / 1
    half                                                 0 / 0        (n/a)
    float                                                1 / 1
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    32, Little-Endian
  Global memory size                              8573157376 (7.984GiB)
  Error Correction support                        No
  Max memory allocation                           2143289344 (1.996GiB)
  Unified memory for Host and Device              No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       32768 bits (4096 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 2048 images
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      10ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            134217728 (128MiB)
  Built-in kernels
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_APPLE_command_queue_priority cl_APPLE_command_queue_select_compute_units cl_khr_fp64

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Apple
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [P0]
  clCreateContext(NULL, ...) [default]            Success [P0]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  Success (1)
    Platform Name                                 Apple
    Device Name                                   Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (2)
    Platform Name                                 Apple
    Device Name                                   Intel(R) UHD Graphics 630
    Device Name                                   AMD Radeon Pro 5500M Compute Engine
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  <checkNullCtxFromType:2238: create context from type CL_DEVICE_TYPE_CUSTOM : error -30>
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (3)
    Platform Name                                 Apple
    Device Name                                   Intel(R) UHD Graphics 630
    Device Name                                   AMD Radeon Pro 5500M Compute Engine
    Device Name                                   Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz

```

</details>

<details>
<summary> `hashcat -I` </summary>

```
hashcat (v6.2.6-775-ga46c9cb81+) starting in backend information mode

Metal Info:
===========

Metal.Version.: 212.8

Backend Device ID #1
  Type...........: GPU
  Vendor.ID......: 2
  Vendor.........: Apple
  Name...........: Intel(R) UHD Graphics 630
  Processor(s)...: 8
  Clock..........: N/A
  Memory.Total...: 1536 MB (limited to 1024 MB allocatable in one block)
  Memory.Free....: 704 MB
  Local.Memory...: 32 KB
  Phys.Location..: built-in
  Registry.ID....: 1846
  Max.TX.Rate....: N/A
  GPU.Properties.: headless 0, low-power 1, removable 0

Backend Device ID #2
  Type...........: GPU
  Vendor.ID......: 2
  Vendor.........: Apple
  Name...........: AMD Radeon Pro 5500M
  Processor(s)...: 8
  Clock..........: N/A
  Memory.Total...: 8176 MB (limited to 3584 MB allocatable in one block)
  Memory.Free....: 8064 MB
  Local.Memory...: 32 KB
  Phys.Location..: built-in
  Registry.ID....: 1924
  Max.TX.Rate....: N/A
  GPU.Properties.: headless 0, low-power 0, removable 0

OpenCL Info:
============

OpenCL Platform ID #1
  Vendor..: Apple
  Name....: Apple
  Version.: OpenCL 1.2 (Jun  8 2020 17:36:15)

  Backend Device ID #3
    Type...........: CPU
    Vendor.ID......: 8
    Vendor.........: Intel
    Name...........: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Version........: OpenCL 1.2
    Processor(s)...: 16
    Clock..........: 2400
    Memory.Total...: 65536 MB (limited to 8192 MB allocatable in one block)
    Memory.Free....: 32736 MB
    Local.Memory...: 32 KB
    OpenCL.Version.: OpenCL C 1.2
    Driver.Version.: 1.1

  Backend Device ID #4
    Type...........: GPU
    Vendor.ID......: 8
    Vendor.........: Intel Inc.
    Name...........: Intel(R) UHD Graphics 630
    Version........: OpenCL 1.2
    Processor(s)...: 24
    Clock..........: 1250
    Memory.Total...: 1536 MB (limited to 192 MB allocatable in one block)
    Memory.Free....: 704 MB
    Local.Memory...: 64 KB
    OpenCL.Version.: OpenCL C 1.2
    Driver.Version.: 1.2(Oct 29 2020 23:08:45)

  Backend Device ID #5
    Type...........: GPU
    Vendor.ID......: 1
    Vendor.........: AMD
    Name...........: AMD Radeon Pro 5500M Compute Engine
    Version........: OpenCL 1.2
    Processor(s)...: 24
    Clock..........: 1450
    Memory.Total...: 8176 MB (limited to 2044 MB allocatable in one block)
    Memory.Free....: 1920 MB
    Local.Memory...: 32 KB
    OpenCL.Version.: OpenCL C 1.2
    Driver.Version.: 1.2 (Oct 29 2020 23:01:05)

```

</details>

```
On Linux: post lspci output

N/A
```

**Additional context**
The master branch lists Apple OpenCL drivers as unreliable and won't use it unless you give it `--force`, so I expected that Metal backend should work instead.

Also tried on v6.2.6 tag, but I can't even use the Metal API as only Apple silicon machines are supported by default. If I hack the code to enable it, I get the same compile error. I'm not sure if it's because of the system being to old or the differences between Intel and Apple silicon for now (although I think the latter is highly not probable).

**Comment by matrix** [comment]
Hi,
"Intel(R) UHD Graphics 630" is not supported.
Try -d2

**Comment by HarukaMa** [comment]
> Hi,
> "Intel(R) UHD Graphics 630" is not supported.
> Try -d2

I have tried and it didn't work. Anyway, the error message doesn't really look like that is the issue, right?

**Comment by matrix** [comment]
> > Hi,
> > "Intel(R) UHD Graphics 630" is not supported.
> > Try -d2
> 
> I have tried and it didn't work. Anyway, the error message doesn't really look like that is the issue, right?

So the problem is that your gpu cards are not supported by Metal. Is not solvable on our side, it is a limitation of the framework

**Comment by HarukaMa** [comment]
I don't quite understand why the error is "file not found" if the problem is the gpu is not supported by the framework (which I assure you that Metal is of course supported by the Radeon one, why would Apple sell hardwares without Metal support?).

You might want to give a better explanation than "gpu not supported by metal" as either that's not a full explanation or the reason is wrong.

**Comment by matrix** [comment]
> I don't quite understand why the error is "file not found" if the problem is the gpu is not supported by the framework (which I assure you that Metal is of course supported by the Radeon one, why would Apple sell hardwares without Metal support?).
> 
> You might want to give a better explanation than "gpu not supported by metal" as either that's not a full explanation or the reason is wrong.

while developing this feature I realized that it didn't work on my old macbook because of the same problem you are describing. This happens both when the version of metal is too old and when the hardware is old. Unfortunately, I could not find a workaround; it is a "feature" of Metal. 
Also if you compare for example the value of "Clock" you get using Metal and OpenCL (hashcat -I output) you will see for yourself that there is an inconsistency. Since the 2.x version of Metal runs on new hardware, we have tried to force the Clock value to 8 in order to perform calculations preparatory to device initialization, but sometimes (as in your case) this is not enough.
I hope the explanation is now exhaustive
